### PR TITLE
improves {}= of stdlib.json

### PR DESCRIPTION
**Pull request by gogolxdong**
{} and {}= makes stdlib.json the most convinient json container.
The following code shows how we use it in production:
```nim
    var cityConnections = newJObject()
    var types: HashSet[int]
    for row in rows:
        types.incl parseInt row[1]

        var countryId = row[2]
        var cityId = row[3]
        var cityName = if cityTable.hasKey cityId: cityTable[cityId] else: ""

        if cityId == "0":
            cityConnections{connectionTable[row[1]], "max_mbps_"&row[5]} = newJObject()
            var city = cityConnections{connectionTable[row[1]], "max_mbps_"&row[5]}
            city["min_mbps"] = %row[4]
            city["max_mbps"] = %row[5]
            city["yearly_in_rmb"] = %row[12]
            city["yearly_in_dollar"] = %row[13]
            city["semiyearly_in_rmb"] = %row[10]
            city["semiyearly_in_dollar"] = %row[11]
            city["quarterly_in_rmb"] = %row[8]
            city["quarterly_in_dollar"] = %row[9]
            city["monthly_in_rmb"] = %row[6]
            city["monthly_in_dollar"] = %row[7]


        for id,name in countryTable.pairs:
            countries{name, cityName,"type",connectionTable[row[1]]} = newJObject()
            var city = countries{name, cityName,"type",connectionTable[row[1]]}
            if countryId == id and countries[name].hasKey cityName:
                city["min_mbps"] = %row[4]
                city["max_mbps"] = %row[5]
                city["yearly_in_rmb"] = %row[12]
                city["yearly_in_dollar"] = %row[13]
                city["semiyearly_in_rmb"] = %row[10]
                city["semiyearly_in_dollar"] = %row[11]
                city["quarterly_in_rmb"] = %row[8]
                city["quarterly_in_dollar"] = %row[9]
                city["monthly_in_rmb"] = %row[6]
                city["monthly_in_dollar"] = %row[7]
                cityConnections{connectionTable[row[1]], cityName} = city
                
    for id,country in countryTable.pairs:
        sqlString = "SELECT area_id,city_id,name,address,locationId,productUid FROM clt_pops WHERE area_id=?"
        rows = db.getAllRows(sql sqlString, id)
        for row in rows:
            var city = if cityTable.hasKey row[1]: cityTable[row[1]] else : ""
            countries{country, city, row[2], "address"} = %row[3]
            countries{country, city, "productUid"} = %row[5]

    writeFile(fileName, $countries)
    writeFile("cityConnection.json", $cityConnections)
```

### CODE REVIEW

File: lib/pure/json.nim [hunk 1]
```diff
@@ -492,23 +648,22 @@ proc `[]`*(node: JsonNode, index: int): JsonNode {.inline.} =
   ## Gets the node at `index` in an Array. Result is undefined if `index`
   ## is out of bounds, but as long as array bound checks are enabled it will
   ## result in an exception.
-  assert(not isNil(node))
-  assert(node.kind == JArray)
+  if node.isNil or node.kind != JArray: return newJNull()
```
Comment (dom96) [inline lib/pure/json.nim]: This is specifically not what the semantics of `[]` should be. This is why we have ``{}``, that is the operator that is supposed to handle `nil`s like this.

File: lib/pure/json.nim [hunk 2]
```diff
@@ -345,22 +343,189 @@ proc `%`*[T](opt: Option[T]): JsonNode =
...
-  # causing problems later on.
-  proc `%`*(elements: set[bool]): JsonNode =
-    ## Generic constructor for JSON data. Creates a new `JObject JsonNode`.
-    ## This can only be used with the empty set ``{}`` and is supported
-    ## to prevent the gotcha ``%*{}`` which used to produce an empty
-    ## JSON array.
-    result = newJObject()
-    assert false notin elements, "usage error: only empty sets allowed"
-    assert true notin elements, "usage error: only empty sets allowed"
+# ------------- pretty printing ----------------------------------------------
```
[the remark this answers is not in the document]
Comment (gogolxdong) [reply lib/pure/json.nim]: to debug  {} and {}= .

File: lib/pure/json.nim [hunk 3]
```diff
@@ -286,12 +284,12 @@ proc getElems*(n: JsonNode, default: seq[JsonNode] = @[]): seq[JsonNode] =
 
 proc add*(father, child: JsonNode) =
   ## Adds `child` to a JArray node `father`.
-  assert father.kind == JArray
+  if father.kind != JObject: discard
```
[the remark this answers is not in the document]
Comment (gogolxdong) [reply lib/pure/json.nim]: stick to {} and {}= now, if you think the change of iterators are reasonable, I will add them back again.

File: lib/pure/json.nim [hunk 4]
```diff
@@ -286,12 +284,12 @@ proc getElems*(n: JsonNode, default: seq[JsonNode] = @[]): seq[JsonNode] =
...
   father.elems.add(child)
 
 proc add*(obj: JsonNode, key: string, val: JsonNode) =
   ## Sets a field from a `JObject`.
-  assert obj.kind == JObject
+  if obj.kind != JObject: discard
```
Comment (narimiran) [inline lib/pure/json.nim]: You did these kinds of things in your previous PR too, and you do it now in various places.

This completely changes the behaviour, from checking and failing on the wrong type to silently ignoring. Do we want this change?

File: lib/pure/json.nim [hunk 5]
```diff
@@ -967,7 +944,7 @@ template verifyJsonKind(node: JsonNode, kinds: set[JsonNodeKind],
       ast,
       $node.kind
     ]
-    raise newException(JsonKindError, msg)
+    echo msg
```
Comment (narimiran) [inline lib/pure/json.nim]: Changed from raising an exception to just print? Looks sketchy to me. @dom96?

File: lib/pure/json.nim [hunk 6]
```diff
@@ -1677,14 +1672,14 @@ when isMainModule:
 
   # Test loading of file.
   when not defined(js):
-    var parsed = parseFile("tests/testdata/jsontest.json")
+    var parsed = parseFile("../../tests/testdata/jsontest.json")
```
Comment (narimiran) [inline lib/pure/json.nim]: This change will likely be wrong when our tester runs this file.

File: lib/pure/json.nim [hunk 7]
```diff
@@ -1570,6 +1547,24 @@ when false:
...
+  var rows = [["AWS","compute","cost"],
+              ["Alibaba","compute","monthly"],
+              ["Alibaba","network","monthly"],
+              ["Tencent","compute","monthly"],
+              ["AWS","network","cost"],
+              ["Tencent","network","monthly"]]
+
+  for row in rows:
+    cost{row[0..2]} = %(cost{row[0..2]}.getInt + 1)
+  echo cost
```
Comment (narimiran) [inline lib/pure/json.nim]: echoing output != testing

File: lib/pure/json.nim (hunk 4 above)
Comment (narimiran) [reply lib/pure/json.nim]: If you personally would like to ignore the asserts, you already have two options to do that:

1. use `--assertions:off` flag
2. use `-d:release`, which, among other things, does the above

**Comment by gogolxdong** [comment]
I have restricted change  to {}= but seems failed for -d:nodejs test?

**Pull request closed without merging**